python/paddle/v2/fluid/tests/test_fetch_var.py

```diff
@@ -0,0 +1,23 @@
+import paddle.v2.fluid as fluid
+import paddle.v2.fluid.layers as layers
+import op_test
+import numpy
+import unittest
+
+
+class TestFetchVar(op_test.OpTest):
+    def test_fetch_var(self):
+        val = numpy.array([1, 3, 5]).astype(numpy.int32)
+        x = layers.create_tensor(dtype="int32", persistable=True, name="x")
+        layers.assign(input=val, output=x)
+        exe = fluid.Executor(fluid.CPUPlace())
+        exe.run(fluid.default_main_program(), feed={}, fetch_list=[])
+        fetched_x = fluid.fetch_var("x")
```

Comment:
This is a potential risk that memory optimization transpiler will rename some variable to reuse memory. So, after memory optimization, fetch_var by name may redirect to another variable.

Reply:
But I believe we need to give the user the rights to retrieve the parameter.
Now we have two parts will change the var name, one is backward, one is memory optimization, right?

Reply:
If a variable is persistable, will its name be changed? If so, we can restrict to persistable variables.

Another way is to run a Program using fetch_op to get the variables, but it seems to me it will also have the renaming issue.

Reply:
Also in fluid.io._clone_var_in_block() also depends on the name not changing. How should it be handled?

Reply:
> If a variable is persistable, will its name be changed? If so, we can restrict to persistable variables.

In memory optimization transpiler, the name of persistable variable will not be changed. 
https://github.com/PaddlePaddle/Paddle/blob/e0ae76f222c5dd2484ed0ee9b5d013ff937c5cbb/python/paddle/v2/fluid/memory_optimization_transpiler.py#L139-L141
It seems that _clone_var_in_block also only handles persistable variable
https://github.com/PaddlePaddle/Paddle/blob/e0ae76f222c5dd2484ed0ee9b5d013ff937c5cbb/python/paddle/v2/fluid/io.py#L21-L29

Reply:
I'll change fetch_var to only work with persistable variable.

Reply:
Are all variables in global_scope persistable?

Reply:
executor.run() function has a scope argument. If a variable is created in that scope, will its name be changed?

Reply:
> Are all variables in global_scope persistable?

Not exactly. But all persistable variable(like parameter/accuracy variable) will be created in global scope, and its name will not be changed. 

> executor.run() function has a scope argument. If a variable is created in that scope, will its name be changed?

The scope argument is global scope, and inside the run method, a local scope can be created. Whether a variable name is changed is decided by persistable attribute, not scope.

Reply:
But when can non-persistable variable be created in the scope passing to executor.run()?

Reply:
The persistable variables are including

1. Parameters
2. Learning Rates
3. Momentums and similar variables
4. User fetched variables

The other layer's outputs are non-persistable. They will be created and destroyed every mini-batch.

However, if the layer's intermediate outputs are useful after `executor.run`, we can either

1. Set the `var.persistable = True`
2. Pass `create_local_scope = False`, which is used when implement operators like RNN.

Reply:
Will the name of a variable in the global scope be changed in any situations? If not, then we can use the function a wrote to retrieve its value, probably with more detailed comment and error meeesage

Reply:
I didn't find a way to check a core.Variable object is persistable or not. If there is one, I can simply check it in the function.

Reply:
> I didn't find a way to check a core.Variable object is persistable or not. If there is one, I can simply check it in the function.

Please refer to 
https://github.com/PaddlePaddle/Paddle/blob/03c858a9bee98dbbcbedd7a6be886402a2e97609/python/paddle/v2/fluid/framework.py#L242-L244

Reply:
Python framework.Variable is not core.Variable

Reply:
We do not have core.Variable, we have core.VarDesc. The C++ VarDesc is exposed to python in pybind.cc. Please refer to https://github.com/PaddlePaddle/Paddle/blob/03c858a9bee98dbbcbedd7a6be886402a2e97609/paddle/pybind/pybind.cc#L443-L446

Reply:
scope.find_var() returns core.Variable object. You can use type(var) to see its type. See the following lines of code:
https://github.com/PaddlePaddle/Paddle/blob/03c858a9bee98dbbcbedd7a6be886402a2e97609/paddle/pybind/pybind.cc#L250
https://github.com/PaddlePaddle/Paddle/blob/c3e062e97565e5a790c826022c422c33a95d0b46/paddle/framework/scope.h#L56